extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/SplitPackageProcessor.java

```diff
@@ -28,26 +32,48 @@
     void splitPackageDetection(ApplicationArchivesBuildItem archivesBuildItem,
             // Dummy producer to make sure the build step executes
             BuildProducer<ValidationPhaseBuildItem.ValidationErrorBuildItem> dummy) {
-        Set<String> allKnownPackages = new HashSet<>();
-        Set<String> splitPackages = new HashSet<>();
+        Map<String, Set<ApplicationArchive>> packageToArchiveMap = new HashMap<>();
         // for all app archives
         for (ApplicationArchive archive : archivesBuildItem.getAllApplicationArchives()) {
-            Set<String> singleArchivePackages = new HashSet<>();
             // and for each known class in each archive
             for (ClassInfo classInfo : archive.getIndex().getKnownClasses()) {
-                singleArchivePackages.add(DotNames.packageName(classInfo.name()));
+                String packageName = DotNames.packageName(classInfo.name());
+                packageToArchiveMap.compute(packageName, (key, val) -> {
+                    Set<ApplicationArchive> returnValue = val == null ? new HashSet<>() : val;
+                    returnValue.add(archive);
+                    return returnValue;
+                });
             }
-            // perform intersection
-            Set<String> singleArchiveSplitPackages = new HashSet<>(singleArchivePackages);
-            singleArchiveSplitPackages.retainAll(allKnownPackages);
-            // add single archive packages to set of all known packages
-            allKnownPackages.addAll(singleArchivePackages);
-            // and remember all split packages
-            splitPackages.addAll(singleArchiveSplitPackages);
         }
-        if (!splitPackages.isEmpty()) {
+        // for each split package, create something like
+        // - "com.me.app.sub" found in [archiveA, archiveB]
+        StringBuilder splitPackages = new StringBuilder();
+        for (String packageName : packageToArchiveMap.keySet()) {
+            Set<ApplicationArchive> applicationArchives = packageToArchiveMap.get(packageName);
+            if (applicationArchives.size() > 1) {
+                splitPackages.append("\n- \"" + packageName + "\" found in " + "[");
+                Iterator<ApplicationArchive> iterator = applicationArchives.iterator();
+                while (iterator.hasNext()) {
+                    AppArtifactKey a = iterator.next().getArtifactKey();
+                    if (a == null) {
+                        // can be null for instance in test mode where all application classes go under target/classes
+                        // TODO not sure if this can happen in more cases?
```

Comment:
So it should be easy to detect the root archive if you compare the `ApplicationArchivesBuildItem.getRootArchive()`. And for other archives we could probably just use the first path from the `ApplicationArchiveImpl.getPaths()` collection? And if empty then just output `unknown archive`. @geoand WDYT?

Reply:
Yeah, that sounds reasonable to me

Reply:
Good point with root archive but you still need to somehow represent it in the output and I am not sure saying "current project" would give user much enlightenment :D

> just use the first path from the ApplicationArchiveImpl.getPaths()

How is the first path most reliable and how do you ever get empty paths?

> And for other archives...

Well, I found the group and artifact IDs to be easily understandable. Or do you think otherwise?

Reply:
> Good point with root archive but you still need to somehow represent it in the output and I am not sure saying "current project" would give user much enlightenment :D

`target/classes` is OK, except it does not make sense for gradle users ;-). Anything like "root application archive" or "application classes" is IMO good enough.

> Well, I found the group and artifact IDs to be easily understandable. Or do you think otherwise?

Of course, if it's available then just use it! I was just talking about the branch where `a == null`...

> How is the first path most reliable and how do you ever get empty paths?

I'm not quite sure TBH :-) but it seems that for most of the archives the first path is the "archive location".

Reply:
@mkouba Updated with suggestions (and squashed commits). The detection is now way more fine grained and from from group/artifact ID to `application classes` with a fallback to paths and uses `uknown archive` as a last resort.